### PR TITLE
Flexible CSRs

### PR DESCRIPTION
**Pull request by piotro888**
Additions to the CSR register interface

### CODE REVIEW

File: coreblocks/priv/csr/csr_register.py [hunk 1]
```diff
@@ -75,40 +54,88 @@ class CSRRegister(Elaboratable):
...
+        )
+        self.fu_write_filter = MethodFilter(
+            self.fu_write_map.method, lambda tm, ms: fu_write_filtermap(tm, ms["data"])[0]
+        )
+        self.fu_read_map = MethodMap(
+            self._internal_fu_read,
+            o_transform=(csr_layouts._fu_read, lambda tm, ms: {"data": fu_read_map(tm, ms["data"])}),
+        )
+
+        # Methods connected autatically by CSRUnit
```
Comment (xThaid) [inline coreblocks/priv/csr/csr_register.py:125]: typo

File: test/structs_common/test_csr.py [hunk 2]
```diff
@@ -262,4 +262,99 @@ def test_randomized(self):
...
+                expected = input
+                if input & 2:
+                    expected += 3
+
+                expected &= ~(2**32)
+
+                expected <<= 1
+                expected &= 2**34 - 1
+
+            print(input, expected, output)
```
Comment (xThaid) [inline test/structs_common/test_csr.py:286]: print left after debugging

**Review by xThaid — APPROVED** [review]
looks good, just some minor comments

File: coreblocks/func_blocks/csr/csr.py [hunk 3]
```diff
@@ -68,11 +91,11 @@ def __init__(self, gen_params: GenParams):
         self.regfile: dict[int, tuple[Method, Method]] = {}
 
     def _create_regfile(self):
-        # Fills `self.regfile` with `CSRRegister`s provided by `CSRListKey` depenecy.
+        # Fills `self.regfile` with `CSRRegister`s provided by `CSRListKey` dependency.
         for csr in self.dependency_manager.get_dependency(CSRListKey()):
             if csr.csr_number in self.regfile:
                 raise RuntimeError(f"CSR number {csr.csr_number} already registered")
-            self.regfile[csr.csr_number] = (csr._fu_read, csr._fu_write)
+            self.regfile[csr.csr_number] = (csr._fu_read, csr._fu_write)  # type: ignore
```
Comment (tilk) [inline coreblocks/func_blocks/csr/csr.py:98]: Why? Is it because the `csr_number` is optional now? If so, an assert should be here.

Comment (piotro888) [reply coreblocks/func_blocks/csr/csr.py]: good point

**Review by tilk — APPROVED** [review]

**Pull request merged**